store/store.go

```diff
@@ -244,10 +269,34 @@ func generateMongoQuery(p *Params) bson.M {
 
 		if len(andQuery) > 0 {
 			groupDataQuery["$and"] = andQuery
+			finalQuery = groupDataQuery
+		} else if skipParamsQuery || len(p.LevelFilter) == 0 {
+			finalQuery = groupDataQuery
+		} else {
+			paramQuery := []bson.M{}
+			// create the level filter as string
+			levelFilterAsString := []string{}
+			for value := range p.LevelFilter {
+				levelFilterAsString = append(levelFilterAsString, strconv.Itoa(value))
+			}
+
+			// split the group in 2 queries for deviceParameters
+			orQuery := bson.M{}
+			for key, value := range groupDataQuery {
+				orQuery[key] = value
+			}
+			groupDataQuery["subType"] = bson.M{"$ne": "deviceParameter"}
+			orQuery["type"] = "deviceEvent"
+			orQuery["subType"] = "deviceParameter"
+			orQuery["level"] = bson.M{"$in": levelFilterAsString}
+
+			paramQuery = append(paramQuery, groupDataQuery)
+			paramQuery = append(paramQuery, orQuery)
+			finalQuery = bson.M{"$or": paramQuery}
```

Comment:
On my small dataset with the following index created : 
```
db.deviceData.createIndex( { _userId : 1, _active : 1, _schemaVersion : 1, time : -1 }, {name: "UserIdTypeWeighted"})
```
Performance is improved, with if we transform the query from : 
  * ```$or: [deviceParametersQuery,otherDataQuery]```
  * to : ```$and: [commonFilters,$or :[deviceParameterSpecificFilters,otherDataSpecificFilters] ]```
 
Examples of querys : 
```
// current version 
db.getCollection('deviceData').find({
	$or:[
		{
			_active:true, 
			_schemaVersion:{$gte:1, $lte:99}, 
			_userId:"b5d7d3da8b",
			time:{$gte:"2020-01-13T00:00:00Z", $lte:"2020-03-10T20:16:41.185Z"},
                        level:{$in:[0, 1]}, 
                        subType:"deviceParameter", 
                        type:"deviceEvent"
		},
		{
                    _active:true, 
                    _schemaVersion:{$gte:1, $lte:99}, 
                    _userId:"b5d7d3da8b",
                    time:{$gte:"2020-01-13T00:00:00Z", $lte:"2020-03-10T20:16:41.185Z"},
                    subType:{ $ne: "deviceParameter" }, 
                    type:{ $ne: "deviceEvent" }
		}
	]
})
// new version
db.getCollection('deviceData').find({
	$and:[
		{
			_active:true, 
			_schemaVersion:{$gte:1, $lte:99}, 
			_userId:"b5d7d3da8b",
			time:{$gte:"2020-01-13T00:00:00Z", $lte:"2020-03-10T20:16:41.185Z"}
		},
		{
			$or:[
				{ 
					level:{$in:[0, 1]}, 
					subType:"deviceParameter", 
					type:"deviceEvent"
				}, 
				{
                                    subType:{ $ne: "deviceParameter" },
                                    type:{ $ne: "deviceEvent" }
                                    
				}
			]
		}
	]
})
```
The execution time drops from 138ms to 56ms (34ms if we don't filter levels) 

```suggestion
			paramQuery = append(paramQuery, groupDataQuery)
			
			deviceParametersQuery := bson.M{}
			deviceParametersQuery["type"] = "deviceEvent"
			deviceParametersQuery["subType"] = "deviceParameter"
			deviceParametersQuery["level"] = bson.M{"$in": levelFilterAsString}

			otherDataQuery := bson.M{}
			otherDataQuery["type"] = bson.M{"$ne": "deviceEvent"}
			otherDataQuery["subType"] = bson.M{"$ne": "deviceParameter"}
			
			orQuery := []bson.M{}
			orQuery = append(orQuery, deviceParametersQuery)
			orQuery = append(orQuery, otherDataQuery)
			
			paramQuery = append(paramQuery, bson.M{"$or": orQuery})

			finalQuery = bson.M{"$and": paramQuery}
```

Reply:
resolved in 6c6352304311f4b3994d9361c105bbc26dbf3d29